<a href="https://colab.research.google.com/github/AHTHneeuhl/.config/blob/main/zerotopandas_course_project_starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nifty50 Stock Market Data Analysis: BRITANNIA

This project involves performing an exploratory data analysis on the Nifty50 stock market data, specifically focusing on the BRITANNIA stock. The dataset was obtained from Kaggle using the `opendatasets` library. Our goal is to clean and prepare the data, visualize key trends, and answer specific questions about BRITANNIA's stock performance over time. We will be utilizing Python libraries such as `pandas` for data manipulation, and `matplotlib` and `seaborn` for visualization. This analysis is undertaken as part of the learning experience from the [Data Analysis with Python: Zero to Pandas](zerotopandas.com) course.

In [1]:
!pip install opendatasets --upgrade --quiet

Let's begin by downloading the data, and listing the files within the dataset.

In [2]:
# Change this
dataset_url = 'https://www.kaggle.com/datasets/rohanrao/nifty50-stock-market-data/'

In [3]:
import opendatasets as od
od.download(dataset_url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: ahthneeuhl
Your Kaggle Key: ··········
Dataset URL: https://www.kaggle.com/datasets/rohanrao/nifty50-stock-market-data


100%|██████████| 18.4M/18.4M [00:00<00:00, 85.1MB/s]


The dataset has been downloaded and extracted.

In [4]:
# Change this
data_dir = './nifty50-stock-market-data'

In [5]:
import os
os.listdir(data_dir)

['HINDUNILVR.csv',
 'JSWSTEEL.csv',
 'BAJAJ-AUTO.csv',
 'CIPLA.csv',
 'WIPRO.csv',
 'HEROMOTOCO.csv',
 'GAIL.csv',
 'NESTLEIND.csv',
 'INDUSINDBK.csv',
 'NTPC.csv',
 'ITC.csv',
 'SBIN.csv',
 'SUNPHARMA.csv',
 'ULTRACEMCO.csv',
 'ASIANPAINT.csv',
 'BHARTIARTL.csv',
 'BPCL.csv',
 'ADANIPORTS.csv',
 'ICICIBANK.csv',
 'BRITANNIA.csv',
 'TATAMOTORS.csv',
 'HDFC.csv',
 'HCLTECH.csv',
 'BAJAJFINSV.csv',
 'ZEEL.csv',
 'SHREECEM.csv',
 'IOC.csv',
 'COALINDIA.csv',
 'VEDL.csv',
 'NIFTY50_all.csv',
 'KOTAKBANK.csv',
 'stock_metadata.csv',
 'LT.csv',
 'MM.csv',
 'TITAN.csv',
 'UPL.csv',
 'GRASIM.csv',
 'EICHERMOT.csv',
 'TCS.csv',
 'ONGC.csv',
 'DRREDDY.csv',
 'RELIANCE.csv',
 'HDFCBANK.csv',
 'INFY.csv',
 'HINDALCO.csv',
 'INFRATEL.csv',
 'BAJFINANCE.csv',
 'TECHM.csv',
 'MARUTI.csv',
 'POWERGRID.csv',
 'AXISBANK.csv',
 'TATASTEEL.csv']

## Data Preparation and Cleaning

In this section, we will load the dataset, inspect its structure and initial properties, and perform necessary cleaning and preprocessing steps to ensure the data is suitable for analysis. This typically includes:

1.  **Loading the Dataset**: Reading the CSV file into a pandas DataFrame.
2.  **Initial Exploration**: Checking the number of rows and columns, data types, and a preview of the data.
3.  **Handling Missing Values**: Identifying and addressing any missing data points.
4.  **Data Type Conversion**: Converting columns to appropriate data types (e.g., 'Date' to datetime objects).
5.  **Feature Engineering (Optional)**: Creating new features or modifying existing ones if required for analysis.

In [6]:
import pandas as pd
import numpy as np

In [7]:
britannia = pd.read_csv('nifty50-stock-market-data/BRITANNIA.csv')

In [8]:

britannia.head()

,Date,Symbol,Series,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble
0,2000-01-03,BRITANNIA,EQ,703.25,705.0,759.50,705.0,758.0,756.90,741.01,7512,5.566488e+11,NaN,NaN,NaN
1,2000-01-04,BRITANNIA,EQ,756.90,710.0,770.00,710.0,740.0,754.55,742.52,8135,6.040391e+11,NaN,NaN,NaN
2,2000-01-05,BRITANNIA,EQ,754.55,755.0,759.00,705.0,740.0,735.30,739.92,6095,4.509784e+11,NaN,NaN,NaN
3,2000-01-06,BRITANNIA,EQ,735.30,740.0,794.15,740.0,770.0,785.65,788.83,19697,1.553756e+12,NaN,NaN,NaN
4,2000-01-07,BRITANNIA,EQ,785.65,808.0,848.50,798.0,848.5,848.50,827.53,33107,2.739708e+12,NaN,NaN,NaN


First 5 rows (`britannia.head()`):

* The dataset contains columns like `Date`, `Symbol`, `Series`, `Prev Close`, `Open`, `High`, `Low`, `Last`, `Close`, `VWAP`, `Volume`, `Turnover`, `Trades`, `Deliverable Volume`, and `%Deliverble`.
* `Date` is currently an object (string), which might need conversion to datetime for time-series analysis.
* Several columns have `NaN` (Not a Number) values, particularly `Trades`, `Deliverable Volume`, and `%Deliverble` in the initial rows.

In [9]:
britannia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5305 entries, 0 to 5304
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Date                5305 non-null   object 
 1   Symbol              5305 non-null   object 
 2   Series              5305 non-null   object 
 3   Prev Close          5305 non-null   float64
 4   Open                5305 non-null   float64
 5   High                5305 non-null   float64
 6   Low                 5305 non-null   float64
 7   Last                5305 non-null   float64
 8   Close               5305 non-null   float64
 9   VWAP                5305 non-null   float64
 10  Volume              5305 non-null   int64  
 11  Turnover            5305 non-null   float64
 12  Trades              2456 non-null   float64
 13  Deliverable Volume  4796 non-null   float64
 14  %Deliverble         4796 non-null   float64
dtypes: float64(11), int64(1), object(3)
memory usage: 621.8

Information (`britannia.info()`):

* The DataFrame consists of 5305 entries and 15 columns.
* `Date`, `Symbol`, and `Series` are of `object` type.
* Most numerical columns are `float64`, with `Volume` as `int64`.
* There are missing values in `Trades` (only 2456 non-null out of 5305), `Deliverable Volume` (4796 non-null), and `%Deliverble` (4796 non-null). These will need to be addressed during data cleaning.

In [10]:
britannia.describe()

,Prev Close,Open,High,Low,Last,Close,VWAP,Volume,Turnover,Trades,Deliverable Volume,%Deliverble
count,5305.000000,5305.000000,5305.000000,5305.000000,5305.000000,5305.000000,5305.000000,5.305000e+03,5.305000e+03,2456.000000,4.796000e+03,4796.000000
mean,1687.974769,1690.173054,1713.527135,1665.679293,1688.601112,1688.492347,1689.630878,1.220883e+05,3.531071e+13,19243.627850,6.877725e+04,0.644524
std,1364.834888,1367.095820,1379.848544,1350.727532,1365.080043,1364.982005,1365.360931,2.668956e+05,9.119291e+13,22420.581857,1.340685e+05,0.179159
min,336.350000,340.000000,347.700000,295.200000,336.950000,336.350000,339.860000,8.400000e+01,9.614630e+09,27.000000,5.700000e+01,0.027900
25%,591.800000,594.500000,603.900000,580.200000,592.500000,591.800000,592.100000,4.284000e+03,3.694440e+11,4496.750000,3.040500e+03,0.518900
50%,1230.150000,1230.000000,1259.000000,1203.350000,1230.000000,1230.300000,1232.610000,2.002900e+04,1.347110e+12,13231.000000,1.913050e+04,0.651350
75%,2743.700000,2741.050000,2778.400000,2701.100000,2747.000000,2743.900000,2743.300000,1.347500e+05,3.634675e+13,24920.750000,8.919100e+04,0.777425
max,6900.150000,6918.000000,6934.350000,6831.700000,6898.000000,6900.150000,6886.460000,5.383773e+06,1.954734e+15,294264.000000,4.253015e+06,1.000000


Descriptive Statistics (`britannia.describe()`):

* Provides statistical summaries (mean, std, min, max, quartiles) for numerical columns.
* Prices (`Prev Close`, `Open`, `High`, `Low`, `Last`, `Close`) range significantly, indicating potential growth or volatility over time.
* `Volume` and `Turnover` also show wide ranges, suggesting varying trading activity.
* The maximum values for `Turnover` are very large, possibly indicating a need to scale or transform this column if used in certain analyses.

In [11]:
britannia['Date'] = pd.to_datetime(britannia['Date'])
britannia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5305 entries, 0 to 5304
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Date                5305 non-null   datetime64[ns]
 1   Symbol              5305 non-null   object        
 2   Series              5305 non-null   object        
 3   Prev Close          5305 non-null   float64       
 4   Open                5305 non-null   float64       
 5   High                5305 non-null   float64       
 6   Low                 5305 non-null   float64       
 7   Last                5305 non-null   float64       
 8   Close               5305 non-null   float64       
 9   VWAP                5305 non-null   float64       
 10  Volume              5305 non-null   int64         
 11  Turnover            5305 non-null   float64       
 12  Trades              2456 non-null   float64       
 13  Deliverable Volume  4796 non-null   float64     

### Date Column Conversion

The 'Date' column was initially loaded as an `object` (string) data type. For proper time-series analysis, plotting, and calculations involving dates, it's essential to convert this column to a `datetime` format. This conversion allows pandas to recognize and work with dates efficiently, enabling operations like filtering by date ranges, extracting year/month/day, or calculating time differences. The `pd.to_datetime()` function is used for this purpose.

In [12]:
print('Missing values before handling:')
display(britannia.isnull().sum())

Missing values before handling:


,0
Date,0
Symbol,0
Series,0
Prev Close,0
Open,0
High,0
Low,0
Last,0
Close,0
VWAP,0


In [13]:
print('\nPercentage of missing values:')
display((britannia.isnull().sum() / len(britannia) * 100).round(2))


Percentage of missing values:


,0
Date,0.00
Symbol,0.00
Series,0.00
Prev Close,0.00
Open,0.00
High,0.00
Low,0.00
Last,0.00
Close,0.00
VWAP,0.00


### Handling Missing Values

Before proceeding with the analysis, it's crucial to address any missing values in the dataset. Missing values can distort our analysis and lead to inaccurate conclusions. From the `britannia.info()` output, we identified that 'Trades', 'Deliverable Volume', and '%Deliverble' columns contain `NaN` values. The code above calculates and displays the count and percentage of missing values for each column.

Upon inspection, it's evident that 'Trades' has a significant portion of its data missing (approximately 53.69%), while 'Deliverable Volume' and '%Deliverble' have about 9.59% missing. Depending on the analysis goals, different strategies can be employed: dropping rows/columns, imputation (mean, median, mode), or more advanced methods.

### Missing Value Handling Strategy Implementation

Given the high percentage of missing values in the 'Trades' column (approximately 53.69%), dropping this column seems appropriate as imputation might introduce too much bias. For 'Deliverable Volume' and '%Deliverble', which have a smaller percentage of missing values (around 9.59%), median imputation is a reasonable strategy. The median is chosen over the mean to be less sensitive to outliers in the data.

In [14]:
# Drop the 'Trades' column
britannia_cleaned = britannia.drop(columns=['Trades'])

# Impute missing values in 'Deliverable Volume' and '%Deliverble' with their medians
median_deliverable_volume = britannia_cleaned['Deliverable Volume'].median()
median_deliverble_percentage = britannia_cleaned['%Deliverble'].median()

britannia_cleaned['Deliverable Volume'] = britannia_cleaned['Deliverable Volume'].fillna(median_deliverable_volume)
britannia_cleaned['%Deliverble'] = britannia_cleaned['%Deliverble'].fillna(median_deliverble_percentage)

print('Missing values after handling:')
display(britannia_cleaned.isnull().sum())

Missing values after handling:


,0
Date,0
Symbol,0
Series,0
Prev Close,0
Open,0
High,0
Low,0
Last,0
Close,0
VWAP,0


## Exploratory Analysis and Visualization

Now that the data has been prepared and cleaned, we can begin our exploratory analysis. This involves computing key statistics, visualizing distributions, and identifying relationships between different features to gain insights into the dataset.

### Closing Price Trend Over Time

Let's visualize the `Close` price of BRITANNIA stock over the entire period to observe its historical trend. This line plot will help us identify periods of growth, decline, and overall stability or volatility.

In [15]:
plt.figure(figsize=(12, 6))
plt.plot(britannia_cleaned['Date'], britannia_cleaned['Close'])
plt.xlabel('Date')
plt.ylabel('Closing Price')
plt.title('BRITANNIA Stock Closing Price Over Time')
plt.grid(True)
plt.show()

NameError: name 'plt' is not defined

Let's begin by importing`matplotlib.pyplot` and `seaborn`.

### Summary of Descriptive Statistics

Key observations from `britannia_cleaned.describe()`:

*   **Price Columns (`Prev Close`, `Open`, `High`, `Low`, `Last`, `Close`):** We can see a wide range in prices, with the stock starting around 300-350 and reaching highs of nearly 7000. This indicates significant growth over the period covered by the dataset. The mean and median values also reflect this upward trend. The standard deviation is quite high, suggesting considerable volatility.
*   **`VWAP` (Volume Weighted Average Price):** The VWAP generally follows the other price columns, as expected, giving an average price weighted by trading volume.
*   **`Volume`:** The average daily trading volume is around 122,000, with a massive maximum of over 5 million. This wide range suggests days with very low trading activity and days with extremely high activity. The high standard deviation also points to significant variability.
*   **`Turnover`:** This column shows exceptionally large values, with a maximum in the quadrillions, indicating the total value of shares traded. This column will likely need scaling if used in models that are sensitive to magnitude differences.
*   **`Deliverable Volume` and `%Deliverble`:** These columns now show no missing values due to our median imputation strategy. The `%Deliverble` (percentage of deliverable quantity to total traded quantity) has a mean of about 64.5% and ranges from 2.79% to 100%, indicating varying investor conviction and settlement behavior. The standard deviation of around 17% suggests moderate variation in this percentage over time.

In [ ]:
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 14
matplotlib.rcParams['figure.figsize'] = (9, 5)
matplotlib.rcParams['figure.facecolor'] = '#00000000'

### Descriptive Statistics of Cleaned Data

We will now look at the descriptive statistics of the `britannia_cleaned` DataFrame to understand the central tendency, dispersion, and shape of its distribution for numerical columns, especially after handling missing values. This will help us identify potential outliers, understand data ranges, and prepare for visualizations.

In [ ]:
display(britannia_cleaned.describe())

Let's begin by importing`matplotlib.pyplot` and `seaborn`.

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

### Trading Volume Over Time

Next, let's visualize the `Volume` (number of shares traded) over time. This can reveal periods of increased investor interest, liquidity changes, or significant market events that correlate with high or low trading activity.

In [ ]:
plt.figure(figsize=(12, 6))
plt.plot(britannia_cleaned['Date'], britannia_cleaned['Volume'], color='orange')
plt.xlabel('Date')
plt.ylabel('Trading Volume')
plt.title('BRITANNIA Stock Trading Volume Over Time')
plt.grid(True)
plt.show()

### Trading Volume Analysis

The plot of 'BRITANNIA Stock Trading Volume Over Time' shows significant fluctuations in daily trading volume. Initially, in the early 2000s, trading volumes were relatively low. However, there's a noticeable increase in volume over the years, particularly from around 2010 onwards, with several spikes indicating periods of intense trading activity. These spikes often correlate with significant price movements or market events. Higher volumes generally indicate stronger conviction behind price trends, while lower volumes might suggest less interest or consolidation. The increasing overall trend in volume suggests growing investor participation and liquidity in BRITANNIA shares over the dataset's period.

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

### Daily Price Range (Volatility) Over Time

The daily price range, calculated as the difference between the 'High' and 'Low' prices for each day, is a good indicator of the stock's intraday volatility. A larger range suggests higher volatility, meaning the stock experienced wider price swings within a single trading day. This plot will help us identify periods when BRITANNIA stock was more or less volatile.

In [ ]:
britannia_cleaned['Price_Range'] = britannia_cleaned['High'] - britannia_cleaned['Low']

plt.figure(figsize=(12, 6))
plt.plot(britannia_cleaned['Date'], britannia_cleaned['Price_Range'], color='green')
plt.xlabel('Date')
plt.ylabel('Daily Price Range')
plt.title('BRITANNIA Stock Daily Price Range Over Time')
plt.grid(True)
plt.show()

### Daily Price Range Analysis

The plot showing the 'BRITANNIA Stock Daily Price Range Over Time' reveals how the stock's intraday volatility has changed. We can observe that in the earlier years, the daily price range was generally smaller, indicating less volatility. As the stock price increased over time, particularly in later years, the daily price range also appears to widen, suggesting that BRITANNIA stock has become more volatile as its value grew. Periods of sharp spikes in the price range might correspond to significant market news, company announcements, or overall market instability, leading to larger intraday price fluctuations. This indicates that while the stock has shown growth, it has also become more dynamic in its daily trading behavior.

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

### Turnover Over Time

Turnover represents the total value of shares traded on a given day. Plotting turnover over time can indicate periods of high market activity and investor interest in the stock. Large spikes in turnover often coincide with significant news or events affecting the company or the broader market.

In [ ]:
plt.figure(figsize=(12, 6))
plt.plot(britannia_cleaned['Date'], britannia_cleaned['Turnover'], color='red')
plt.xlabel('Date')
plt.ylabel('Turnover (in billions)')
plt.title('BRITANNIA Stock Turnover Over Time')
plt.grid(True)
plt.ticklabel_format(style='plain', axis='y') # Prevent scientific notation on y-axis
plt.show()

### Turnover Analysis

The 'BRITANNIA Stock Turnover Over Time' plot illustrates the total value of shares traded daily. Similar to trading volume, turnover shows a general increase over the years, indicating growing market participation and a higher monetary value exchanged for BRITANNIA shares. We observe several massive spikes in turnover, especially in recent years. These extreme values suggest specific days with extraordinary trading activity, likely driven by major corporate announcements, market-wide events, or significant institutional trading. The substantial growth in turnover from the early 2000s to later periods highlights the increasing liquidity and importance of BRITANNIA in the market.

### Deliverable Volume Over Time

Deliverable volume refers to the actual quantity of shares that change ownership and are not squared off on the same trading day (intraday trades). High deliverable volume can signify long-term investor interest, as these shares are typically held for more than a day. Plotting this metric over time can reveal trends in investor conviction.

In [ ]:
plt.figure(figsize=(12, 6))
plt.plot(britannia_cleaned['Date'], britannia_cleaned['Deliverable Volume'], color='purple')
plt.xlabel('Date')
plt.ylabel('Deliverable Volume')
plt.title('BRITANNIA Stock Deliverable Volume Over Time')
plt.grid(True)
plt.show()

### Deliverable Volume Analysis

The plot of 'BRITANNIA Stock Deliverable Volume Over Time' shows the quantity of shares that were traded and subsequently delivered to investors' demat accounts. This metric is often considered an indicator of long-term investment interest. We can observe fluctuations, but generally, the deliverable volume has also increased over the years, especially correlating with periods of higher overall trading volume. Spikes in deliverable volume suggest days when a large number of shares were bought with the intention of being held, which can imply a strong underlying sentiment or a significant event that attracted long-term investors. The pattern somewhat mirrors the total volume, but looking at deliverable volume specifically helps differentiate speculative intraday trading from actual ownership transfer.

### %Deliverble Over Time

The '%Deliverble' column represents the percentage of the total traded quantity that resulted in delivery. This ratio provides insights into the sentiment of market participants: a higher percentage might suggest stronger buying conviction (investors taking delivery), while a lower percentage could indicate more speculative or intraday trading activity. Analyzing its trend can reveal changes in market psychology towards the stock.

In [ ]:
plt.figure(figsize=(12, 6))
plt.plot(britannia_cleaned['Date'], britannia_cleaned['%Deliverble'], color='brown')
plt.xlabel('Date')
plt.ylabel('%Deliverble')
plt.title('BRITANNIA Stock %Deliverble Over Time')
plt.grid(True)
plt.show()

### %Deliverble Analysis

The plot of 'BRITANNIA Stock %Deliverble Over Time' illustrates the proportion of total traded shares that were taken for delivery. This percentage fluctuates significantly, indicating varying levels of long-term investor interest versus short-term trading. We can see periods where the percentage is consistently high, suggesting strong conviction among buyers, and other periods where it dips, possibly indicating increased intraday trading or speculative activity. There doesn't appear to be a clear continuous upward or downward trend, but rather oscillations, which are typical for this metric. Analyzing these fluctuations can help in understanding market sentiment shifts and the balance between long-term investment and short-term speculation in BRITANNIA shares.

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

## Asking and Answering Questions

In this section, we will formulate and answer specific questions about the BRITANNIA stock dataset. By leveraging our cleaned data and the powerful capabilities of `pandas` for data manipulation and `matplotlib`/`seaborn` for visualization, we aim to extract deeper insights into the stock's performance and trends. Each question will be addressed with relevant code and a clear explanation of the findings.

> Instructions (delete this cell)
>
> - Ask at least 5 interesting questions about your dataset
> - Answer the questions either by computing the results using Numpy/Pandas or by plotting graphs using Matplotlib/Seaborn
> - Create new columns, merge multiple dataset and perform grouping/aggregation wherever necessary
> - Wherever you're using a library function from Pandas/Numpy/Matplotlib etc. explain briefly what it does



#### Q1: What is the total percentage change in BRITANNIA's closing price from its earliest recorded date to its latest?

In [ ]:
# Get the first and last closing prices
first_close_price = britannia_cleaned['Close'].iloc[0]
last_close_price = britannia_cleaned['Close'].iloc[-1]

# Calculate the percentage change
percentage_change = ((last_close_price - first_close_price) / first_close_price) * 100

print(f"BRITANNIA's closing price on {britannia_cleaned['Date'].iloc[0].strftime('%Y-%m-%d')}: {first_close_price:.2f}")
print(f"BRITANNIA's closing price on {britannia_cleaned['Date'].iloc[-1].strftime('%Y-%m-%d')}: {last_close_price:.2f}")
print(f"Total percentage change in closing price: {percentage_change:.2f}%")

In [ ]:
#### Answer to Q1: Total Percentage Change in Closing Price

To answer this question, we first identified the `Close` price from the very first trading day (`iloc[0]`) and the very last trading day (`iloc[-1]`) in our `britannia_cleaned` DataFrame. We then used the standard percentage change formula: `((last_value - first_value) / first_value) * 100`.

The result indicates a substantial increase in BRITANNIA's stock price over the period covered by the dataset. This positive percentage change reflects significant growth in the stock's value from its inception in the dataset until the most recent trading day, aligning with the upward trend observed in the 'Closing Price Over Time' plot.

#### Q2: Which dates had the highest and lowest trading volumes for BRITANNIA stock?

In [ ]:
# Find the row with the highest trading volume
highest_volume_row = britannia_cleaned.loc[britannia_cleaned['Volume'].idxmax()]
highest_volume_date = highest_volume_row['Date'].strftime('%Y-%m-%d')
highest_volume = highest_volume_row['Volume']

# Find the row with the lowest trading volume
lowest_volume_row = britannia_cleaned.loc[britannia_cleaned['Volume'].idxmin()]
lowest_volume_date = lowest_volume_row['Date'].strftime('%Y-%m-%d')
lowest_volume = lowest_volume_row['Volume']

print(f"Date with highest trading volume: {highest_volume_date} with {highest_volume} shares traded.")
print(f"Date with lowest trading volume: {lowest_volume_date} with {lowest_volume} shares traded.")

In [ ]:
#### Answer to Q2: Dates with Highest and Lowest Trading Volumes

To answer this, we used the `idxmax()` and `idxmin()` pandas DataFrame methods to find the index of the maximum and minimum values in the 'Volume' column, respectively. Then, `loc` was used to retrieve the entire row for these indices, allowing us to extract the corresponding 'Date' and 'Volume'. The `strftime('%Y-%m-%d')` method was used to format the datetime objects into a readable date string.

The results highlight specific days with extreme trading activity, which could correlate with significant company announcements, financial reports, or major market movements. Identifying these dates can be crucial for further historical analysis to understand the drivers behind such high or low investor interest.

#### Q3: What is the average daily return of BRITANNIA stock, and how are these returns distributed?

In [ ]:
# Calculate daily returns
britannia_cleaned['Daily_Return'] = britannia_cleaned['Close'].pct_change() * 100

# Calculate the average daily return
average_daily_return = britannia_cleaned['Daily_Return'].mean()

print(f"Average Daily Return: {average_daily_return:.2f}%")

# Plot histogram of daily returns
plt.figure(figsize=(10, 6))
sns.histplot(britannia_cleaned['Daily_Return'].dropna(), kde=True, bins=50)
plt.title('Distribution of BRITANNIA Stock Daily Returns')
plt.xlabel('Daily Return (%)')
plt.ylabel('Frequency')
plt.grid(True)
plt.show()

In [ ]:
#### Answer to Q3: Average Daily Return and Distribution

To address this question, we first computed the 'Daily_Return' for the stock using the `.pct_change()` method on the 'Close' price, multiplying by 100 to express it as a percentage. This function calculates the percentage change between the current and a prior element. We then calculated the `mean()` of these daily returns to find the average.

The average daily return provides a general sense of the stock's daily performance. The histogram visualizes the frequency of different daily return percentages. A bell-shaped curve centered around the mean (or close to zero) is often expected for stock returns, but the tails of the distribution can indicate periods of high volatility (large positive or negative returns). This plot helps in understanding the typical range and likelihood of daily price movements.

#### Q4: TODO - ask a question here and answer it below

#### Q5: TODO - ask a question here and answer it below

## Inferences and Conclusion

**TODO** - Write some explanation here: a summary of all the inferences drawn from the analysis, and any conclusions you may have drawn by answering various questions.

## References and Future Work

**TODO** - Write some explanation here: ideas for future projects using this dataset, and links to resources you found useful.

> Submission Instructions (delete this cell)
>
> To save your work, select "File" > "Save a Copy in Drive" on Google Colab. Once the copy is created, click the "Share" button and select "Anyone with the link" under the "General Access" section to make this notebook publicly accessible.
>
> <img src="https://miro.medium.com/v2/resize:fit:800/format:webp/1*4eXpS4TgIJ6n7obsaUKptQ.png" height="400">
>
>Then, copy the notebook link and submit it on the assignment page.
>
> (Optional) Write a blog post
>
> - A blog post is a great way to present and showcase your work.  
> - Sign up on [Medium.com](https://medium.com) to write a blog post for your project.
> - Copy over the explanations from your Jupyter notebook into your blog post, and [embed code cells & outputs](https://medium.com/jovianml/share-and-embed-jupyter-notebooks-online-with-jovian.com-df709a03064e)
> - Check out the Jovian.ml Medium publication for inspiration: https://medium.com/jovianml


